### DISCUSSION TITLE
SQLAlchemy 2.0 Core: is the engine and its pool thread-safe?

### DISCUSSION THREAD

**Discussion by LucaMaurelliC** [sqlalchemy/sqlalchemy#9992] [Usage Questions]
I'm new to SQLAlchemy and to Python concurrent.futures, so bear with me please. I have read the documentation at https://docs.sqlalchemy.org/en/20/core/pooling.html#using-connection-pools-with-multiprocessing-or-os-fork but it refers to multi-processing so far, and I'm confused whether it applies to threads or not also.

In my application I'm using the library gRPC to implement some async APIs between clients and a server, the listening server is set-up using a PoolThreadExecutor of 10 workers using the concurrent.futures library. If I wish to setup SQL Alchemy right, so that the engine and its connection works as expected with concurrent threads, what should I do?

To date I create an engine object globally, started the server, and every time the server receives a request, it spins a thread and perform a function asyncroniously from the receiving client request.

```
def serve():
    ... some code ...
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    ... some code ...
    server.start()
    server.wait_for_termination()


if __name__ == "__main__":
    ... some code ...
    engine = create_engine(connection_url)
    serve()
```
Is this configuration right and safe?

**Comment by zzzeek** [comment]
the engine is thread safe yes.  individual Connection objects are not.   we try to describe this at https://docs.sqlalchemy.org/en/20/core/connections.html#basic-usage
